store/generated_key/v2/eigenda.go

```diff
@@ -141,12 +149,59 @@ func (e Store) BackendType() common.BackendType {
 //
 // Since v2 methods for fetching a payload are responsible for verifying the received bytes against the certificate,
 // this Verify method only needs to check the cert on chain. That is why the third parameter is ignored.
-func (e Store) Verify(ctx context.Context, certBytes []byte, _ []byte) error {
+func (e Store) Verify(ctx context.Context, certBytes []byte, _ []byte, opts common.VerifyOpts) error {
 	var eigenDACert coretypes.EigenDACert
 	err := rlp.DecodeBytes(certBytes, &eigenDACert)
 	if err != nil {
 		return fmt.Errorf("RLP decoding EigenDA v2 cert: %w", err)
 	}
+	err = e.verifyRBNRecencyCheck(eigenDACert.BatchHeader.ReferenceBlockNumber, opts)
+	if err != nil {
+		return fmt.Errorf("punctuality check failed: %w", err)
+	}
 
 	return e.verifier.VerifyCertV2(ctx, &eigenDACert)
 }
+
+// Certs in the rollup batcher-inbox that do not respect the below equation are discarded.
+//
+//	cert.RBN < cert.L1InclusionBlock <= cert.RBN + RBNRecencyWindowSize
+//
+// where ReferenceBlockNumber (RBN) is the block number at which operator stakes are referenced
+// to verify the signature thresholds, and L1InclusionBlock is the block at which the cert
+// was included in the L1 block.
+//
+// This check serves 2 purposes:
+//  1. liveness: prevents derivation pipeline from stalling on blobs that are no longer availeble on the DA layer
+//  2. safety: prevents a malicious EigenDA sequencer from using a very stale RBN whose operator distribution
+//     does not represent the actual stake distribution. Operators that withdrew a lot of stake would
+//     not be slashable anymore, even though because of the old RBN their signature would count for a lot of stake.
+//
+// Note that for a secure integration, this same check needs to be verified onchain.
+// There are 2 approaches to doing this:
+//  1. Pessimistic approach: use a smart batcher inbox to dissalow stale blobs from even beign included
+//     in the batcher inbox (see https://github.com/ethereum-optimism/design-docs/pull/229)
+//  2. Optimistic approach: verify the check in op-program or hokulea (kona)'s derivation pipeline. See
+//
+// https://github.com/Layr-Labs/hokulea/blob/8c4c89bc4f35d56a3cec2220575a9681d987105c/crates/eigenda/src/eigenda.rs#L90
+func (e Store) verifyRBNRecencyCheck(certRBN uint32, opts common.VerifyOpts) error {
```

Comment:
should this check be moved to the client itself? cc @litt3

Reply:
Ahh ok I wasn't aware that our v2 clients were not doing verification. We synced offline, and think we should at smoe point create a `CertVerifyingPayloadRetrieverClient` which would do everything proxy is doing (validate cert, then proceed to query blob from relay + nodes if needed).

At that point we could move this verification logic to that client in a single swoop cleanup. Agree?

Reply:
Yeah, this sounds good to me

Reply:
[df30640](https://github.com/Layr-Labs/eigenda-proxy/pull/378/commits/df30640ccda39ea3584abae02a387a7c20993db0)